File: go/helpers/parser.go
package listings

import (
	"bytes"
	"encoding/csv"
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

// ExtractListing parses the provided HTML string and returns
// extracted listing data as a CSV-formatted string.
// example:
// price;address;link;badge
// $11,000;130 Spring Mountain Dr, Zion Grove, PA 17985;https://www.zillow.com/homedetails/...;On Zillow for 30 days
func ExtractListings(html string) (string, int, error) {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return "", 0, err
	}

	var records [][]string
	records = append(records, []string{
		"price",
		"address",
		"link",
		"badge",
	})

	// Each property card
	doc.Find(".photo-cards > li").Each(func(_ int, card *goquery.Selection) {

		// Badge
		f := func(i int, s *goquery.Selection) bool {
			link, _ := s.Attr("class")
			return strings.HasPrefix(link, "StyledPropertyCardBadge-")
		}

		badge := strings.TrimSpace(
			card.Find("span").FilterFunction(f).First().Text(),
		)
		//fmt.Printf("full badge text: %s\n", badge)

		// Price (first text containing $)
		price := ""
		card.Find("a > div > span").EachWithBreak(func(_ int, s *goquery.Selection) bool {
			t := strings.TrimSpace(s.Text())
			if strings.Contains(t, "$") {
				price = t
				return false
			}
			return true
		})
		//fmt.Printf("price: %s\n", price)

		// Address
		address := strings.TrimSpace(
			card.Find("address, [data-testid='address'], a").First().Text(),
		)
		//fmt.Printf("address: %s\n", address)

		// Deep link
		link := ""
		if href, exists := card.Find("a[href]").First().Attr("href"); exists {
			link = strings.TrimSpace(href)
		}
		//fmt.Printf("link: %s\n", link)

		// Only emit rows that actually look like listings
		if price != "" && address != "" && link != "" {
			records = append(records, []string{
				price,
				address,
				link,
				badge,
			})
			//		} else {
			//			fmt.Println("skip: ", price, address, link, badge)
		}
		//os.Exit(0)
	})

	nav := doc.Find(".search-pagination").Find("span").First().Text()
	nextPage, err := processNavigation(nav)
	if err != nil {
		fmt.Println("warn: ", err)
	}
	var buf bytes.Buffer
	writer := csv.NewWriter(&buf)
	writer.Comma = ';'
	writer.WriteAll(records[1:]) // skip headers

	if err := writer.Error(); err != nil {
		return "", 0, err
	}

	return buf.String(), nextPage, nil
}

func processNavigation(nav string) (int, error) {
	re, err := regexp.Compile(`Page (\w+) of (\w+)`)
	if err != nil {
		return 0, err
	}
	return -1, nil // @TODO remove
	matches := re.FindStringSubmatch(nav)
	first := matches[1]
	second := matches[2]
	if first != second {
		i, err := strconv.Atoi(first)
		if err != nil {
			return 0, err
		}
		i++
		return i, nil
	}
	return -1, err
}

func GetHeaders() (string, error) {
	var records [][]string
	records = append(records, []string{
		"price",
		"address",
		"link",
		"badge",
	})

	var buf bytes.Buffer
	writer := csv.NewWriter(&buf)
	writer.Comma = ';'
	writer.WriteAll(records)

	if err := writer.Error(); err != nil {
		return "", err
	}

	return buf.String(), nil
}

File: parser.py
import csv
import io
import re
from bs4 import BeautifulSoup


def extract_listings(html: str) -> tuple[str, int, Exception | None]:
    """
    Parses the provided HTML string and returns:
    - CSV-formatted string (semicolon-delimited, no headers)
    - next page number (or -1)
    - error (None on success)
    """
    try:
        soup = BeautifulSoup(html, "html.parser")
    except Exception as e:
        return "", 0, e

    records = []

    # Each property card
    for card in soup.select(".photo-cards > li"):

        # Badge: span whose class starts with "StyledPropertyCardBadge-"
        badge = ""
        for span in card.find_all("span"):
            classes = span.get("class", [])
            if any(cls.startswith("StyledPropertyCardBadge-") for cls in classes):
                badge = span.get_text(strip=True)
                break

        # Price: first span text containing $
        price = ""
        for span in card.select("a > div > span"):
            text = span.get_text(strip=True)
            if "$" in text:
                price = text
                break

        # Address
        address_elem = (
            card.find("address")
            or card.find(attrs={"data-testid": "address"})
            or card.find("a")
        )
        address = address_elem.get_text(strip=True) if address_elem else ""

        # Deep link
        link = ""
        a_tag = card.find("a", href=True)
        if a_tag:
            link = a_tag["href"].strip()

        # Only emit valid listings
        if price and address and link:
            records.append([
                price,
                address,
                link,
                badge,
            ])

    # Pagination
    nav_text = ""
    nav_span = soup.select_one(".search-pagination span")
    if nav_span:
        nav_text = nav_span.get_text(strip=True)

    try:
        next_page = process_navigation(nav_text)
    except Exception as e:
        print("warn:", e)
        next_page = -1

    # Write CSV (semicolon-delimited, no headers)
    buf = io.StringIO()
    writer = csv.writer(buf, delimiter=";")
    writer.writerows(records)

    return buf.getvalue(), next_page, None


def process_navigation(nav: str) -> int:
    re_page = re.compile(r"Page (\w+) of (\w+)")
    # Matches the Go behavior: TODO path always returns -1
    return -1

    # Unreachable logic (kept for parity)
    matches = re_page.search(nav)
    if not matches:
        return -1

    first, second = matches.group(1), matches.group(2)
    if first != second:
        return int(first) + 1

    return -1


def get_headers() -> tuple[str, Exception | None]:
    headers = [["price", "address", "link", "badge"]]

    buf = io.StringIO()
    writer = csv.writer(buf, delimiter=";")
    writer.writerows(headers)

    return buf.getvalue(), None
